src/topology/config/mod.rs

```diff
@@ -29,6 +30,63 @@ impl GlobalOptions {
             data_dir: config.data_dir.clone(),
         }
     }
+
+    /// Resolve the `data_dir` option in either the global or local
+    /// config, and validate that it exists and is writable.
+    pub fn resolve_and_validate_data_dir(
+        &self,
+        local_data_dir: Option<&PathBuf>,
+    ) -> Result<PathBuf, String> {
+        let data_dir = local_data_dir
+            .or(self.data_dir.as_ref())
+            .ok_or_else(|| {
+                String::from("data_dir option required, but not given here or globally")
+            })?
+            .to_path_buf();
+        if !data_dir.exists() {
+            return Err(format!(
+                "data_dir '{}' does not exist",
+                data_dir.to_string_lossy()
+            ));
+        }
+        let readonly = std::fs::metadata(&data_dir)
+            .map(|meta| meta.permissions().readonly())
+            .unwrap_or(true);
```

Comment:
this doesn't make much sense if the metadata call fails we can still readonly, wouldn't this indicate that maybe the file/dir doesn't exist?

Reply:
That was moved from the `file` source as-is. The check for the directory existing happens immediately above this, so this can only fail if `stat` fails on an existing directory, which is almost certainly also a permissions error (or kernel bug). I can add an additional error case if you think it makes sense.